File: 06-DevOps/04-Kubernetes/kube-proxy-IPVS负载均衡/go/ipvs_lb.go
// ipvs_lb.go — kube-proxy IPVS 模式负载均衡调度算法模拟
//
// 权威来源:
//   - kubernetes.io/docs/reference/networking/virtual-ips (IPVS 模式 + 11 个调度算法)
//   - kubernetes.io/blog/2018/07/09/ipvs-based-in-cluster-load-balancing-deep-dive
//
// 实现的调度算法:rr / lc / sh / dh / sed / nq / mh (Maglev 简化版)
package main

import "fmt"

const mhN = 137 // 质数,lookup table size = M * N

type Backend struct {
	IP            string
	Port          int
	Weight        int
	ActiveConn    int
	TotalSelected int
}

type Service struct {
	Name      string
	Backends  []Backend
	RRCursor  int
	Lookup    *MaglevTable
}

type MaglevTable struct {
	size   int
	lookup []int
}

func (t *MaglevTable) Pick(srcIP string) int {
	pos := hashIP(srcIP) % t.size
	for t.lookup[pos] == -1 {
		pos = (pos + 1) % t.size
	}
	return t.lookup[pos]
}

func NewMaglevTable(bs []Backend) *MaglevTable {
	n := len(bs)
	size := n * mhN
	t := &MaglevTable{size: size, lookup: make([]int, size)}
	for i := range t.lookup {
		t.lookup[i] = -1
	}
	for i, b := range bs {
		for k := 0; k < mhN; k++ {
			key := fmt.Sprintf("%s/%d/%d", b.IP, i, k)
			pos := hashIP(key) % size
			for t.lookup[pos] != -1 {
				pos = (pos + 1) % size
			}
			t.lookup[pos] = i
		}
	}
	return t
}

// FNV-1a 32-bit
func hashIP(s string) uint32 {
	h := uint32(2166136261)
	for _, ch := range []byte(s) {
		h ^= uint32(ch)
		h *= 16777619
	}
	return h
}

// rr
func lbRR(svc *Service) int {
	idx := svc.RRCursor % len(svc.Backends)
	svc.RRCursor++
	return idx
}

// lc
func lbLC(svc *Service) int {
	best := 0
	for i := 1; i < len(svc.Backends); i++ {
		if svc.Backends[i].ActiveConn < svc.Backends[best].ActiveConn {
			best = i
		}
	}
	return best
}

// sh
func lbSH(svc *Service, srcIP string) int {
	return int(hashIP(srcIP)) % len(svc.Backends)
}

// dh
func lbDH(svc *Service, dstIP string) int {
	return int(hashIP(dstIP)) % len(svc.Backends)
}

// sed: min (C+1)/U
func lbSED(svc *Service) int {
	best := 0
	bestV := float64(svc.Backends[0].ActiveConn+1) / float64(svc.Backends[0].Weight)
	for i := 1; i < len(svc.Backends); i++ {
		v := float64(svc.Backends[i].ActiveConn+1) / float64(svc.Backends[i].Weight)
		if v < bestV {
			best = i
			bestV = v
		}
	}
	return best
}

// nq
func lbNQ(svc *Service) int {
	for i, b := range svc.Backends {
		if b.ActiveConn == 0 {
			return i
		}
	}
	return lbSED(svc)
}

// ============== Demo 1: rr balance ==============
func demo1RRBalance() {
	fmt.Println("\n========== Demo 1: Round Robin distribution over 100 requests ==========")
	svc := &Service{
		Name:     "demo1-svc",
		Backends: []Backend{{IP: "10.244.0.1"}, {IP: "10.244.0.2"}, {IP: "10.244.0.3"}},
	}
	for i := 0; i < 100; i++ {
		idx := lbRR(svc)
		svc.Backends[idx].TotalSelected++
	}
	for _, b := range svc.Backends {
		fmt.Printf("  backend %s: selected %d times\n", b.IP, b.TotalSelected)
	}
}

// ============== Demo 2: all algorithms ==============
func demo2AlgorithmComparison() {
	fmt.Println("\n========== Demo 2: 11 algorithms comparison ==========")
	svc := &Service{
		Name:     "demo2-svc",
		Backends: []Backend{{IP: "10.244.0.1"}, {IP: "10.244.0.2"}, {IP: "10.244.0.3"}},
	}
	srcIPs := []string{"192.168.1.10", "192.168.1.20", "192.168.1.30"}

	fmt.Println("rr (no src affinity):")
	for i := 0; i < 6; i++ {
		fmt.Printf("  → backend %d\n", lbRR(svc))
	}

	fmt.Println("sh (source hash, 3 src):")
	for _, s := range srcIPs {
		idx := lbSH(svc, s)
		fmt.Printf("  src=%s → backend %s\n", s, svc.Backends[idx].IP)
	}

	fmt.Println("dh (dest hash, same dst → same backend):")
	dst := "10.0.0.1"
	for _, s := range srcIPs {
		idx := lbDH(svc, dst)
		fmt.Printf("  src=%s, dst=%s → backend %s\n", s, dst, svc.Backends[idx].IP)
	}

	for i := 0; i < 100; i++ {
		idx := lbRR(svc)
		svc.Backends[idx].ActiveConn++
	}
	fmt.Printf("after 100 rr requests, active_conn = [%d, %d, %d]\n",
		svc.Backends[0].ActiveConn, svc.Backends[1].ActiveConn, svc.Backends[2].ActiveConn)
	fmt.Printf("lc next pick → backend %s (active_conn=%d)\n",
		svc.Backends[lbLC(svc)].IP, svc.Backends[lbLC(svc)].ActiveConn)
	fmt.Printf("sed next pick → backend %s\n", svc.Backends[lbSED(svc)].IP)
	fmt.Printf("nq next pick → backend %s (prefer active_conn=0, fallback sed)\n",
		svc.Backends[lbNQ(svc)].IP)

	fmt.Println("mh (Maglev, src_hash):")
	mh := NewMaglevTable(svc.Backends)
	for _, s := range srcIPs {
		idx := mh.Pick(s)
		fmt.Printf("  src=%s → backend %s\n", s, svc.Backends[idx].IP)
	}
}

// ============== Demo 3: weighted ==============
func demo3WeightedLB() {
	fmt.Println("\n========== Demo 3: Weighted LB (weight=4,2,1 → ~57%,29%,14%) ==========")
	svc := &Service{
		Name: "demo3-svc",
		Backends: []Backend{
			{IP: "10.244.0.1", Weight: 4},
			{IP: "10.244.0.2", Weight: 2},
			{IP: "10.244.0.3", Weight: 1},
		},
	}
	totalW := 4 + 2 + 1
	cursor := 0
	for i := 0; i < 100; i++ {
		wc := cursor % totalW
		var idx int
		if wc < 4 {
			idx = 0
		} else if wc < 6 {
			idx = 1
		} else {
			idx = 2
		}
		cursor++
		svc.Backends[idx].TotalSelected++
	}
	fmt.Println("wrr over 100 requests:")
	for _, b := range svc.Backends {
		pct := 100.0 * float64(b.TotalSelected) / 100
		fmt.Printf("  backend %s (weight=%d): selected %d times (%.1f%%)\n",
			b.IP, b.Weight, b.TotalSelected, pct)
	}
}

// ============== Demo 4: Session Affinity ==============
func demo4SessionAffinity() {
	fmt.Println("\n========== Demo 4: Session Affinity (persistent per src_ip, 10800s) ==========")
	svc := &Service{
		Name: "demo4-svc",
		Backends: []Backend{
			{IP: "10.244.0.1"}, {IP: "10.244.0.2"}, {IP: "10.244.0.3"},
		},
	}
	mh := NewMaglevTable(svc.Backends)
	src := "192.168.1.10"
	first := mh.Pick(src)
	fmt.Printf("first request from %s → backend %s (hash-decided)\n",
		src, svc.Backends[first].IP)
	for i := 0; i < 5; i++ {
		idx := mh.Pick(src)
		svc.Backends[idx].ActiveConn++
		fmt.Printf("  req %d from %s → backend %s (sticky, active_conn=%d)\n",
			i+1, src, svc.Backends[idx].IP, svc.Backends[idx].ActiveConn)
	}
}

func main() {
	demo1RRBalance()
	demo2AlgorithmComparison()
	demo3WeightedLB()
	demo4SessionAffinity()
}

File: 06-DevOps/04-Kubernetes/kube-proxy-IPVS负载均衡/python/ipvs_lb.py
"""
kube-proxy IPVS mode load balancing schedulers.
Based on kubernetes.io/docs/reference/networking/virtual-ips (11 algorithms)
+ kubernetes.io/blog/2018/07/09/ipvs-based-in-cluster-load-balancing-deep-dive
(topology + session affinity 180min).

Algorithms implemented:
  rr / lc / sh / dh / sed / nq / mh (Maglev Hashing simplified)
"""
from dataclasses import dataclass, field
from typing import List, Optional, Dict


@dataclass
class Backend:
    ip: str
    port: int = 8080
    weight: int = 1
    active_conn: int = 0
    total_selected: int = 0


@dataclass
class Service:
    name: str
    backends: List[Backend] = field(default_factory=list)
    rr_cursor: int = 0
    affinity: Dict[str, "AffinityEntry"] = field(default_factory=dict)


@dataclass
class AffinityEntry:
    backend_idx: int
    expires_at_tick: int  # 10800s = 180min per docs


def hash_ip(ip: str) -> int:
    """FNV-1a 32-bit (for sh/dh/mh)."""
    h = 2166136261
    for ch in ip.encode():
        h ^= ch
        h = (h * 16777619) & 0xFFFFFFFF
    return h


# ============== rr ==============
def lb_rr(svc: Service) -> int:
    idx = svc.rr_cursor % len(svc.backends)
    svc.rr_cursor += 1
    return idx


# ============== lc ==============
def lb_lc(svc: Service) -> int:
    return min(range(len(svc.backends)), key=lambda i: svc.backends[i].active_conn)


# ============== sh ==============
def lb_sh(svc: Service, src_ip: str) -> int:
    return hash_ip(src_ip) % len(svc.backends)


# ============== dh ==============
def lb_dh(svc: Service, dst_ip: str) -> int:
    return hash_ip(dst_ip) % len(svc.backends)


# ============== sed ==============
def lb_sed(svc: Service) -> int:
    return min(range(len(svc.backends)),
               key=lambda i: (svc.backends[i].active_conn + 1) / svc.backends[i].weight)


# ============== nq ==============
def lb_nq(svc: Service) -> int:
    for i, b in enumerate(svc.backends):
        if b.active_conn == 0:
            return i
    return lb_sed(svc)


# ============== mh (Maglev Hashing simplified) ==============
MH_N = 137  # prime; lookup table size = M * N


class MaglevTable:
    def __init__(self, backends: List[Backend]):
        self.n = len(backends)
        self.size = self.n * MH_N
        self.lookup = [-1] * self.size
        for i, b in enumerate(backends):
            for k in range(MH_N):
                pos = hash_ip(f"{b.ip}/{i}/{k}") % self.size
                while self.lookup[pos] != -1:
                    pos = (pos + 1) % self.size
                self.lookup[pos] = i

    def pick(self, src_ip: str) -> int:
        pos = hash_ip(src_ip) % self.size
        while self.lookup[pos] == -1:
            pos = (pos + 1) % self.size
        return self.lookup[pos]


# ============== Demo 1: rr balance ==============
def demo1_rr_balance():
    print("\n========== Demo 1: Round Robin distribution over 100 requests ==========")
    svc = Service(name="demo1-svc", backends=[
        Backend("10.244.0.1"), Backend("10.244.0.2"), Backend("10.244.0.3"),
    ])
    for _ in range(100):
        idx = lb_rr(svc)
        svc.backends[idx].total_selected += 1
    for b in svc.backends:
        print(f"  backend {b.ip}: selected {b.total_selected} times")


# ============== Demo 2: all algorithms ==============
def demo2_algorithm_comparison():
    print("\n========== Demo 2: 11 algorithms comparison (100 requests, varying src) ==========")
    svc = Service(name="demo2-svc", backends=[
        Backend("10.244.0.1"), Backend("10.244.0.2"), Backend("10.244.0.3"),
    ])
    src_ips = ["192.168.1.10", "192.168.1.20", "192.168.1.30"]

    print("rr (no src affinity):")
    for _ in range(6):
        print(f"  → backend {lb_rr(svc)}")

    print("sh (source hash, 3 src):")
    for s in src_ips:
        idx = lb_sh(svc, s)
        print(f"  src={s} → backend {svc.backends[idx].ip}")

    print("dh (dest hash, same dst → same backend regardless of src):")
    dst = "10.0.0.1"
    for s in src_ips:
        idx = lb_dh(svc, dst)
        print(f"  src={s}, dst={dst} → backend {svc.backends[idx].ip}")

    # populate active_conn for lc/sed/nq demo
    for _ in range(100):
        idx = lb_rr(svc)
        svc.backends[idx].active_conn += 1
    conns = [b.active_conn for b in svc.backends]
    print(f"after 100 rr requests, active_conn = {conns}")
    print(f"lc next pick → backend {svc.backends[lb_lc(svc)].ip} (active_conn="
          f"{svc.backends[lb_lc(svc)].active_conn})")
    print(f"sed next pick → backend {svc.backends[lb_sed(svc)].ip}")
    print(f"nq next pick → backend {svc.backends[lb_nq(svc)].ip} "
          f"(prefer active_conn=0, fallback sed)")

    print("mh (Maglev, src_hash):")
    mh = MaglevTable(svc.backends)
    for s in src_ips:
        idx = mh.pick(s)
        print(f"  src={s} → backend {svc.backends[idx].ip}")


# ============== Demo 3: weighted ==============
def demo3_weighted_lb():
    print("\n========== Demo 3: Weighted LB (weight=4,2,1 → ~57%,29%,14%) ==========")
    svc = Service(name="demo3-svc", backends=[
        Backend("10.244.0.1", weight=4),
        Backend("10.244.0.2", weight=2),
        Backend("10.244.0.3", weight=1),
    ])
    total_w = sum(b.weight for b in svc.backends)
    cursor = 0
    for _ in range(100):
        wc = cursor % total_w
        if wc < 4:
            idx = 0
        elif wc < 6:
            idx = 1
        else:
            idx = 2
        cursor += 1
        svc.backends[idx].total_selected += 1
    print("wrr over 100 requests:")
    for b in svc.backends:
        pct = 100.0 * b.total_selected / 100
        print(f"  backend {b.ip} (weight={b.weight}): "
              f"selected {b.total_selected} times ({pct:.1f}%)")


# ============== Demo 4: Session Affinity ==============
def demo4_session_affinity():
    print("\n========== Demo 4: Session Affinity (persistent per src_ip, 10800s) ==========")
    svc = Service(name="demo4-svc", backends=[
        Backend("10.244.0.1"), Backend("10.244.0.2"), Backend("10.244.0.3"),
    ])
    mh = MaglevTable(svc.backends)
    src = "192.168.1.10"
    first = mh.pick(src)
    print(f"first request from {src} → backend {svc.backends[first].ip} (hash-decided)")
    for i in range(5):
        idx = mh.pick(src)
        svc.backends[idx].active_conn += 1
        print(f"  req {i+1} from {src} → backend {svc.backends[idx].ip} "
              f"(sticky, active_conn={svc.backends[idx].active_conn})")


if __name__ == "__main__":
    demo1_rr_balance()
    demo2_algorithm_comparison()
    demo3_weighted_lb()
    demo4_session_affinity()
